irr_blocks_index.go

```diff
@@ -0,0 +1,306 @@
+		}
+		if !foundMatching {
+			return nil
+		}
+	}
+
+	ind := &IrrBlocksIndex{
+		store:               store,
+		bundleSizes:         bundleSizes,
+		loadedUpperBoundary: loadedUpperBoundary,
+		nextBlockRefs:       blockRefs,
+		disordered:          make(map[uint64]*PreprocessedBlock),
+	}
+	if len(ind.nextBlockRefs) == 0 {
+		// ensure we either have at least one blockref or have gone through the whole available ranges
+		ind.loadRangesUntil(0)
+	} else {
+		ind.bumpLoadedUpperIrreversibleBlock()
+	}
+
+	return ind
+
+}
+
+// Reorder tells you which blocks should actually be processed and stores the remaining ones
+// when indexedRangeComplete is true, you should stop your indexed Filesource
+func (s *IrrBlocksIndex) Reorder(blk *PreprocessedBlock) (out []*PreprocessedBlock, indexedRangeComplete bool) {
+
+	if len(s.nextBlockRefs) == 0 {
+		// quickly trigger shutdown of that source with indexedRangeComplete==true
+		return nil, true
+	}
+
+	if blk.ID() == s.nextBlockRefs[0].BlockID {
+		out = append(out, blk)
+	} else {
+		s.disordered[blk.Num()] = blk
+	}
+
+	s.RLock()
+	for i, b := range s.nextBlockRefs {
+		if blk.ID() == b.BlockID {
+			s.nextBlockRefs = removeIndex(s.nextBlockRefs, i)
+			break // ensure we don't reuse index 'i' after removeIndex
+		}
+	}
+	s.RUnlock()
+
+	if len(out) == 0 {
+		return
+	}
+
+	nextBoundary := s.loadedUpperBoundary + 1
+	if len(s.nextBlockRefs) != 0 {
+		nextBoundary = s.nextBlockRefs[0].BlockNum
+	}
+	var reorder []*PreprocessedBlock
+	for i, b := range s.disordered {
+		if i > blk.Num() && i < nextBoundary {
+			reorder = append(reorder, b)
+		}
+	}
+	sort.Slice(reorder, func(i, j int) bool { return reorder[i].Num() < reorder[j].Num() })
+	for _, b := range reorder {
+		out = append(out, b)
+		delete(s.disordered, b.Num())
+	}
+
+	// fill the nextBlockRefs if needed
+	if len(s.nextBlockRefs) == 0 {
+		if len(s.disordered) != 0 {
+			panic("bug in irrBlocksIndex reorder or missing blocks in your store")
+		}
+		s.loadRangesUntil(0)
+		indexedRangeComplete = (len(s.nextBlockRefs) == 0)
+	}
+
+	return
+}
+
+// multi-threaded
+func (s *IrrBlocksIndex) Skip(blk BlockRef) bool {
+	if !s.withinIndexRange(blk.Num()) {
+		return true
+	}
+
+	s.RLock()
+	defer s.RUnlock()
+	for _, b := range s.nextBlockRefs {
+		if blk.ID() == b.BlockID {
+			return false
+		}
+	}
+	return true
+
+}
+
+func removeIndex(s []*pbblockmeta.BlockRef, index int) []*pbblockmeta.BlockRef {
+	ret := make([]*pbblockmeta.BlockRef, 0)
+	ret = append(ret, s[:index]...)
+	return append(ret, s[index+1:]...)
+}
+
+// NextBaseBlock additionally includes a lib hasIndex is false so you can bootstrap
+// next source with a forkable
+func (s *IrrBlocksIndex) NextBaseBlock() (baseNum uint64, lib BlockRef, hasIndex bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	var nextWantedBlockNum uint64
+	if len(s.nextBlockRefs) > 0 {
+		hasIndex = true
+		nextWantedBlockNum = s.nextBlockRefs[0].BlockNum
+	} else {
+		nextWantedBlockNum = s.loadedUpperBoundary + 1
```

Comment:
Big smell :) `+ 1` ?:)

Reply:
même chose que l'autre commentaire, +1 is for index boundaries.
index boundaries do not skip any block height, even if no block exists at this height.
This is correct ^_^